### PR TITLE
Check flds for nan

### LINKED ISSUE #1757: Introduce NaN checks for fields passed between components
We want to introduce checks in the coupler that ensure that all fields passed from components are free of NaN values.

There is also value in having some other physical checks (like certain fields that should be non-negative), but feeling from a number of discussions is that those physical checks belong in components, not in the coupler.

### PR DESCRIPTION
**Pull request by jedwards4b**
Checks incoming component fields for NaN in the coupler, fail on first NaN found, report
component name and instance, field name and 1d global index into field.  Add a test in scripts_regression_tests.py that checks this feature is working.  

Test suite: scripts_regression_tests.py , PFS.f19_g17.B1850.cheyenne_intel with and without to measure impact on performance (negligible) 
Test baseline: 
Test namelist changes: 
Test status: bit for bit
Fixes #1757 

User interface changes?: 

Update gh-pages html (Y/N)?:

Code review:

### CODE REVIEW

**Comment by rljacob** [comment]
There's a lot of trailing whitespace differences.  Didn't all the whitespace removal happen a long time ago?

File: src/drivers/mct/main/component_mod.F90 [hunk 1]
```diff
@@ -892,4 +898,41 @@ subroutine component_diag(infodata, comp, flow, comment, info_debug, timer_diag
...
+    integer(IN), pointer :: gpts(:)
+    character(len=CL) :: msg
+    real(r8), pointer :: rattr(:,:)
+
+    lsize = mct_avect_lsize(comp%c2x_cc)
+    nflds = size(comp%c2x_cc%rattr,1)
+    if(lsize <= 1 .and. nflds <= 1) return
+    rattr => comp%c2x_cc%rattr
+    do fld=1,nflds
+       nancnt = 0
```
Comment (billsacks) [inline src/drivers/mct/main/component_mod.F90]: This variable `nancnt` is unused and should be removed

File: src/drivers/mct/main/component_mod.F90 [hunk 2]
```diff
@@ -892,4 +898,41 @@ subroutine component_diag(infodata, comp, flow, comment, info_debug, timer_diag
...
+    integer(in)   :: fld, n            ! iterators
+    integer(IN)  :: rank
+    integer(IN) :: ierr
+    integer(IN), pointer :: gpts(:)
+    character(len=CL) :: msg
+    real(r8), pointer :: rattr(:,:)
+
+    lsize = mct_avect_lsize(comp%c2x_cc)
+    nflds = size(comp%c2x_cc%rattr,1)
+    if(lsize <= 1 .and. nflds <= 1) return
```
Comment (billsacks) [inline src/drivers/mct/main/component_mod.F90]: Why do you return in this case?

**Comment by mvertens** [comment]
@jedwards4b - first thanks for doing this. Also - have we confirmed that the results are bfb in a B1850 test? That would be important to do.

**Comment by billsacks** [comment]
In case you're just following along via email notifications (which don't get updated with edits), I have added a bit to my original review, including an important point (8).

**Comment by billsacks** [comment]
> I'd like this check to be controlled by a flag. I'm concerned it will hurt performance at high resolution.

@rljacob - Would your concerns about performance be alleviated somewhat if the current doubly-nested loop were replaced with an initial check like this, so that shr_infnan_isnan only needs to be called once (noting that it's an elemental function):

```Fortran
if (any(shr_infnan_isnan(rattr))) then
   ! do the existing loop to find the first NaN
end if
```

or are you concerned about the performance of checking individual elements for NaN no matter how it's implemented?

File: src/drivers/mct/main/component_mod.F90 (hunk 2 above)
Comment (jedwards4b) [inline src/drivers/mct/main/component_mod.F90]: The c2x_cc structure is allocated even if it is not used (stub components) in this case size is  1,1 and the field should not be tested.

File: src/drivers/mct/main/component_mod.F90 (hunk 1 above)
Comment (jedwards4b) [inline src/drivers/mct/main/component_mod.F90]: Thanks, I've removed this.

**Comment by jedwards4b** [comment]
For reasons that are not clear to me my final version of scripts_regression_tests.py got overwritten by an older version, rather than fix it again I've pulled those changes and will work on a unit test as @billsacks suggests.   I also agree with @rljacob that this check should be namelist controlled and will add a flag to do so.

File: src/drivers/mct/main/component_mod.F90 (hunk 2 above)
Comment (billsacks) [inline src/drivers/mct/main/component_mod.F90]: Okay. Can you add a comment documenting that? Even better would be if there were a logical in component_type like `c2x_cc_used`, which is set to true in general, but false for stub components, and then you can test against this logical.

**Comment by billsacks** [comment]
@jedwards4b - I have some thoughts this morning about how we could more easily support failure testing like this in unit tests. I'll try implementing this later today.

By the way, I'm fine accepting this PR without tests for now, with the plan to add unit tests later, if you prefer. (I'll be out of the office for about two weeks after today.)

**Comment by jedwards4b** [comment]
@mvertens I have checked bfb against B1850 in alpha07b

**Comment by mvertens** [comment]
@jedwards4b - thanks! I agree with @billsacks that a unit test can be
deferred and this PR should go ahead in order to get it into current CESM
simulations.

**Comment by jedwards4b** [comment]
@rjacob I have added a namelist field nan_check_component_fields it is a logical which defaults to .false. for acme and .true. for cesm.   
@billsacks Unit tests are in place but cannot be turned on until issue #140 is resolved. 

I think that this PR is ready to merge - can you update your reviews if you agree.

File: src/drivers/mct/main/component_mod.F90 [hunk 3]
```diff
@@ -32,7 +32,7 @@ module component_mod
   use mct_mod   ! mct_ wrappers for mct lib
```
Comment (mvertens) [inline src/drivers/mct/main/component_mod.F90]: There still seems to be a lot of white space diffs. Can we remove them?

File: src/drivers/mct/shr/seq_infodata_mod.F90 [hunk 4]
```diff
@@ -276,11 +276,11 @@ MODULE seq_infodata_mod
 !===============================================================================
```
Comment (mvertens) [inline src/drivers/mct/shr/seq_infodata_mod.F90]: Same comment about white space diffs.

File: src/drivers/mct/main/component_mod.F90 (hunk 3 above)
Comment (jedwards4b) [inline src/drivers/mct/main/component_mod.F90]: I have removed all whitespace in #1766 accepting that will remove the differences in this one.

**Comment by billsacks** [comment]
Looks good now - thanks!

**Comment by rljacob** [comment]
Bill asked: "are you concerned about the performance of checking individual elements for NaN no matter how it's implemented?"  Yes.

**Comment by jedwards4b** [comment]
PR #1771 resolves the issue with unit tests in this PR so I have rebased and turned on the unit tests.

**Comment by billsacks** [comment]
Thanks, @jedwards4b !

**Pull request merged**